# Notebook 26 — Structural Features vs SHGAT: Redundancy Analysis

**Date**: 2026-02-28

**Questions**:
1. **H1**: Les embeddings SHGAT encodent-ils la hiérarchie ? (cosine cap ↔ mean children)
2. **H2**: La Jaccard matrix (co-occurrence) est-elle redondante avec la cosine SHGAT ?
3. **H3**: La bigram matrix (transition) est-elle redondante avec la cosine SHGAT ?

Si les embeddings SHGAT capturent déjà l'information structurelle, les features
Jaccard/bigram sont du bruit qui dilue le signal.

In [ ]:
import psycopg2
import numpy as np
import json
import os
import re
from collections import Counter, defaultdict

conn = psycopg2.connect(os.environ.get('DATABASE_URL', 'postgres://casys:Kx9mP2vL7nQ4wRzT@localhost:5432/casys'))
cur = conn.cursor()
print('Connected')

Connected


In [ ]:
# Load tool embeddings: raw BGE-M3 + SHGAT-enriched (both now in DB)
cur.execute("SELECT tool_id, embedding, shgat_embedding FROM tool_embedding ORDER BY tool_id")
tool_rows = cur.fetchall()
tool_ids = [r[0] for r in tool_rows]

tool_embs_raw = {}
tool_embs_shgat = {}
for r in tool_rows:
    raw = json.loads(r[1]) if isinstance(r[1], str) else r[1]
    if raw and len(raw) > 0:
        tool_embs_raw[r[0]] = np.array(raw)
    shgat = json.loads(r[2]) if isinstance(r[2], str) else r[2]
    if shgat and len(shgat) > 0:
        tool_embs_shgat[r[0]] = np.array(shgat)

tool_set = set(tool_ids)
tool_id_to_idx = {t: i for i, t in enumerate(tool_ids)}
print(f'{len(tool_ids)} L0 tools')
print(f'{len(tool_embs_raw)} with raw BGE-M3, {len(tool_embs_shgat)} with SHGAT-enriched')

# Use SHGAT embeddings for analysis (what the GRU actually uses)
embs = tool_embs_shgat if len(tool_embs_shgat) > 0 else tool_embs_raw
emb_type = 'SHGAT-enriched' if embs is tool_embs_shgat else 'BGE-M3 raw'
print(f'Using {emb_type} embeddings for analysis')

920 L0 tools
920 with raw BGE-M3, 920 with SHGAT-enriched
Using SHGAT-enriched embeddings for analysis


In [ ]:
# Load capabilities with tools_used and SHGAT embeddings
def normalize_tool_id(tool_id):
    if not tool_id: return tool_id
    s = tool_id
    for prefix in ['pml.mcp.', 'pml.', 'local.default.', 'local.', 'mcp.']:
        if s.startswith(prefix):
            s = s[len(prefix):]
            break
    if s.startswith('mcp__'): s = s[5:]
    parts = s.split('.')
    if len(parts) >= 2 and ':' not in s:
        return f'{parts[0]}:{parts[1]}'
    return s

cur.execute("""
    SELECT DISTINCT ON (cr.namespace, cr.action)
        cr.namespace || ':' || cr.action as cap_id,
        wp.dag_structure->'tools_used' as tools_used,
        COALESCE(wp.hierarchy_level, 1) as level,
        wp.shgat_embedding,
        wp.intent_embedding
    FROM workflow_pattern wp
    JOIN capability_records cr ON cr.workflow_pattern_id = wp.pattern_id
    WHERE wp.dag_structure->'tools_used' IS NOT NULL
    ORDER BY cr.namespace, cr.action, wp.usage_count DESC
""")
cap_rows = cur.fetchall()

caps = {}  # cap_id -> {children: [], embedding: np.array, level: int}
for row in cap_rows:
    cap_id = row[0]
    raw_tools = row[1] if isinstance(row[1], list) else (json.loads(row[1]) if isinstance(row[1], str) else [])
    children = [normalize_tool_id(t) for t in raw_tools if t]
    children = [c for c in children if c]
    if not children:
        continue
    level = int(row[2])
    # Prefer SHGAT embedding, fallback to intent
    raw_emb = row[3] if row[3] else row[4]
    if not raw_emb:
        continue
    emb = np.array(json.loads(raw_emb) if isinstance(raw_emb, str) else raw_emb)
    if len(emb) == 0:
        continue
    caps[cap_id] = {'children': children, 'embedding': emb, 'level': level}

print(f'{len(caps)} capabilities loaded')
print(f'Levels: {Counter(c["level"] for c in caps.values())}')
with_shgat = sum(1 for row in cap_rows if row[3])
print(f'{with_shgat}/{len(cap_rows)} caps have SHGAT embeddings')

334 capabilities loaded
Levels: Counter({1: 323, 2: 11})
337/337 caps have SHGAT embeddings


## H1: SHGAT Hierarchy Quality

Pour chaque cap, on calcule `cos(cap_emb, mean(children_embs))`.
Si SHGAT encode bien la hiérarchie, les caps devraient être proches de la moyenne de leurs children.

In [ ]:
from numpy.linalg import norm

def cosine_sim(a, b):
    d = norm(a) * norm(b)
    return np.dot(a, b) / d if d > 1e-12 else 0.0

# Flatten caps to L0 children (recursive BFS)
def flatten_to_l0(cap_id, caps_dict, tool_set, visited=None):
    if visited is None:
        visited = set()
    if cap_id in visited:
        return []
    visited.add(cap_id)
    l0 = []
    for child in caps_dict.get(cap_id, {}).get('children', []):
        if child in tool_set:
            l0.append(child)
        elif child in caps_dict:
            l0.extend(flatten_to_l0(child, caps_dict, tool_set, visited))
    return l0

# Compute hierarchy quality for each cap
hierarchy_scores = []
caps_with_emb = 0
caps_no_children = 0

for cap_id, cap in caps.items():
    l0_children = flatten_to_l0(cap_id, caps, tool_set)
    # Get embeddings for L0 children that are in our embedding set
    child_embs = [embs[c] for c in l0_children if c in embs]
    if not child_embs:
        caps_no_children += 1
        continue
    caps_with_emb += 1
    
    # Mean of children embeddings
    mean_child = np.mean(child_embs, axis=0)
    sim = cosine_sim(cap['embedding'], mean_child)
    hierarchy_scores.append({
        'cap_id': cap_id,
        'level': cap['level'],
        'n_children': len(child_embs),
        'cos_sim': sim,
    })

print(f'{caps_with_emb} caps with L0 children in embedding space')
print(f'{caps_no_children} caps with no L0 children in embedding space')

sims = [s['cos_sim'] for s in hierarchy_scores]
print(f'\nCosine(cap, mean_children) distribution:')
print(f'  Mean:   {np.mean(sims):.4f}')
print(f'  Median: {np.median(sims):.4f}')
print(f'  Std:    {np.std(sims):.4f}')
print(f'  Min:    {np.min(sims):.4f}')
print(f'  Max:    {np.max(sims):.4f}')
print(f'  >0.8:   {sum(1 for s in sims if s > 0.8)} ({100*sum(1 for s in sims if s > 0.8)/len(sims):.1f}%)')
print(f'  >0.5:   {sum(1 for s in sims if s > 0.5)} ({100*sum(1 for s in sims if s > 0.5)/len(sims):.1f}%)')
print(f'  <0.3:   {sum(1 for s in sims if s < 0.3)} ({100*sum(1 for s in sims if s < 0.3)/len(sims):.1f}%)')

301 caps with L0 children in embedding space
33 caps with no L0 children in embedding space

Cosine(cap, mean_children) distribution:
  Mean:   0.9379
  Median: 0.9935
  Std:    0.2229
  Min:    -0.0355
  Max:    0.9961
  >0.8:   282 (93.7%)
  >0.5:   285 (94.7%)
  <0.3:   16 (5.3%)


In [ ]:
# Histogram of hierarchy quality scores
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

fig, axes = plt.subplots(1, 2, figsize=(14, 5))

# Overall distribution
axes[0].hist(sims, bins=30, edgecolor='black', alpha=0.7, color='steelblue')
axes[0].axvline(np.mean(sims), color='red', linestyle='--', label=f'Mean={np.mean(sims):.3f}')
axes[0].axvline(np.median(sims), color='orange', linestyle='--', label=f'Median={np.median(sims):.3f}')
axes[0].set_xlabel('Cosine(cap_emb, mean_children_embs)')
axes[0].set_ylabel('Count')
axes[0].set_title('H1: SHGAT Hierarchy Quality')
axes[0].legend()

# By number of children
n_children = [s['n_children'] for s in hierarchy_scores]
axes[1].scatter(n_children, sims, alpha=0.3, s=10)
axes[1].set_xlabel('Number of L0 children')
axes[1].set_ylabel('Cosine(cap, mean_children)')
axes[1].set_title('Hierarchy quality vs cap size')

plt.tight_layout()
plt.savefig('26-shgat-hierarchy-quality.png', dpi=150)
plt.show()
print('Saved: 26-shgat-hierarchy-quality.png')

Saved: 26-shgat-hierarchy-quality.png


## H2: Jaccard Co-occurrence vs SHGAT Cosine

Le Jaccard mesure combien de caps deux tools partagent. SHGAT devrait encoder ça dans les embeddings.
Si `corr(jaccard_ij, cosine_shgat_ij)` est élevé, Jaccard est redondant.

In [ ]:
# Build Jaccard matrix from cap co-occurrence
# For each tool pair (i,j): Jaccard = |caps_i ∩ caps_j| / |caps_i ∪ caps_j|
N = len(tool_ids)

# Build tool → cap membership
tool_to_caps = defaultdict(set)
for cap_id, cap in caps.items():
    l0_children = flatten_to_l0(cap_id, caps, tool_set)
    for t in l0_children:
        if t in tool_id_to_idx:
            tool_to_caps[t].add(cap_id)

tools_with_caps = sum(1 for t in tool_ids if len(tool_to_caps[t]) > 0)
print(f'{tools_with_caps}/{N} tools belong to at least one cap')

# Sample tool pairs for correlation (full matrix is N^2 = 846K, but most are zero)
# Focus on pairs where both tools belong to at least one cap
active_tools = [t for t in tool_ids if len(tool_to_caps[t]) > 0]
print(f'Computing Jaccard for {len(active_tools)} active tools ({len(active_tools)**2} pairs)...')

jaccard_pairs = []
cosine_pairs = []

for i, t1 in enumerate(active_tools):
    if t1 not in embs:
        continue
    caps1 = tool_to_caps[t1]
    for j in range(i+1, len(active_tools)):
        t2 = active_tools[j]
        if t2 not in embs:
            continue
        caps2 = tool_to_caps[t2]
        intersection = len(caps1 & caps2)
        union = len(caps1 | caps2)
        if union == 0:
            continue
        jac = intersection / union
        cos = cosine_sim(embs[t1], embs[t2])
        jaccard_pairs.append(jac)
        cosine_pairs.append(cos)

jaccard_arr = np.array(jaccard_pairs)
cosine_arr = np.array(cosine_pairs)

# Correlation
corr = np.corrcoef(jaccard_arr, cosine_arr)[0, 1]
print(f'\nJaccard vs SHGAT cosine correlation: {corr:.4f}')
print(f'{len(jaccard_pairs)} tool pairs analyzed')
print(f'Jaccard > 0: {sum(1 for j in jaccard_pairs if j > 0)} ({100*sum(1 for j in jaccard_pairs if j > 0)/len(jaccard_pairs):.1f}%)')
print(f'Cosine > 0.5 when Jaccard > 0: {sum(1 for j, c in zip(jaccard_pairs, cosine_pairs) if j > 0 and c > 0.5)}/{sum(1 for j in jaccard_pairs if j > 0)}')

156/920 tools belong to at least one cap
Computing Jaccard for 156 active tools (24336 pairs)...

Jaccard vs SHGAT cosine correlation: 0.2829
12090 tool pairs analyzed
Jaccard > 0: 265 (2.2%)
Cosine > 0.5 when Jaccard > 0: 265/265


In [ ]:
# Scatter plot: Jaccard vs SHGAT cosine
fig, ax = plt.subplots(figsize=(8, 6))

# Color by Jaccard value
ax.scatter(jaccard_arr, cosine_arr, alpha=0.1, s=3, color='steelblue')
ax.set_xlabel('Jaccard (cap co-occurrence)')
ax.set_ylabel('Cosine (SHGAT embedding)')
ax.set_title(f'H2: Jaccard vs SHGAT — corr={corr:.3f}')

# Add trend line
z = np.polyfit(jaccard_arr, cosine_arr, 1)
p = np.poly1d(z)
x_trend = np.linspace(0, max(jaccard_arr), 100)
ax.plot(x_trend, p(x_trend), 'r--', alpha=0.8, label=f'Trend: y={z[0]:.2f}x+{z[1]:.2f}')
ax.legend()

plt.tight_layout()
plt.savefig('26-jaccard-vs-shgat.png', dpi=150)
plt.show()
print('Saved: 26-jaccard-vs-shgat.png')

Saved: 26-jaccard-vs-shgat.png


## H3: Bigram Transitions vs SHGAT Cosine

Le bigram mesure P(tool_j | tool_i) dans les traces. Si les embeddings SHGAT
placent les tools séquentiels proches, le bigram bias est redondant.

In [ ]:
# Build bigram counts from traces
cur.execute("""
    SELECT task_results
    FROM execution_trace
    WHERE task_results IS NOT NULL
      AND jsonb_array_length(task_results) >= 2
    ORDER BY executed_at DESC
    LIMIT 3000
""")
trace_rows = cur.fetchall()
print(f'{len(trace_rows)} traces loaded')

bigram_counts = defaultdict(lambda: defaultdict(int))
total_bigrams = 0

for row in trace_rows:
    results = row[0] if isinstance(row[0], list) else json.loads(row[0])
    tools = []
    for tr in results:
        t = tr.get('tool') or tr.get('toolName') or tr.get('tool_name')
        if t:
            tools.append(normalize_tool_id(t))
    for i in range(len(tools) - 1):
        if tools[i] in tool_id_to_idx and tools[i+1] in tool_id_to_idx:
            bigram_counts[tools[i]][tools[i+1]] += 1
            total_bigrams += 1

print(f'{total_bigrams} bigrams from {len(bigram_counts)} unique first-tools')

# Compare bigram probability with SHGAT cosine
bigram_probs = []
bigram_cosines = []

for t1, successors in bigram_counts.items():
    if t1 not in embs:
        continue
    total = sum(successors.values())
    for t2, count in successors.items():
        if t2 not in embs or t1 == t2:
            continue
        prob = count / total
        cos = cosine_sim(embs[t1], embs[t2])
        bigram_probs.append(prob)
        bigram_cosines.append(cos)

bigram_prob_arr = np.array(bigram_probs)
bigram_cos_arr = np.array(bigram_cosines)

corr_bigram = np.corrcoef(bigram_prob_arr, bigram_cos_arr)[0, 1]
print(f'\nBigram P(j|i) vs SHGAT cosine correlation: {corr_bigram:.4f}')
print(f'{len(bigram_probs)} (tool_i, tool_j) transitions analyzed')

282 traces loaded
584 bigrams from 86 unique first-tools

Bigram P(j|i) vs SHGAT cosine correlation: 0.3358
158 (tool_i, tool_j) transitions analyzed


In [ ]:
# Scatter plot: Bigram P(j|i) vs SHGAT cosine
fig, ax = plt.subplots(figsize=(8, 6))
ax.scatter(bigram_prob_arr, bigram_cos_arr, alpha=0.15, s=5, color='darkorange')
ax.set_xlabel('Bigram P(tool_j | tool_i)')
ax.set_ylabel('Cosine (SHGAT embedding)')
ax.set_title(f'H3: Bigram vs SHGAT — corr={corr_bigram:.3f}')

z = np.polyfit(bigram_prob_arr, bigram_cos_arr, 1)
p = np.poly1d(z)
x_trend = np.linspace(0, max(bigram_prob_arr), 100)
ax.plot(x_trend, p(x_trend), 'r--', alpha=0.8, label=f'Trend: y={z[0]:.2f}x+{z[1]:.2f}')
ax.legend()

plt.tight_layout()
plt.savefig('26-bigram-vs-shgat.png', dpi=150)
plt.show()
print('Saved: 26-bigram-vs-shgat.png')

Saved: 26-bigram-vs-shgat.png


## Summary

| Hypothesis | Metric | Result | Verdict |
|:-----------|:-------|:-------|:--------|
| H1: SHGAT encodes hierarchy | cos(cap, mean_children) | **0.94 mean, 0.99 median** | **OUI** — 93.7% caps > 0.8 |
| H2: Jaccard redundant with SHGAT | corr(jaccard, cosine) | **0.28** | **Partiellement** — corrélation faible |
| H3: Bigram redundant with SHGAT | corr(bigram, cosine) | **0.34** | **Partiellement** — corrélation faible |

### Analyse

**H1 est un résultat fort** : le SHGAT encode excellemment la hiérarchie. 93.7% des caps ont
un cosine > 0.8 avec la moyenne de leurs children. Le similarity head du GRU (`softmax(h · E^T)`)
devrait naturellement scorer les caps haut quand le contexte matche — la hiérarchie est dans les embeddings.

**H2/H3 sont nuancés** : corrélation 0.28-0.34 = Jaccard/bigram capturent de l'info que SHGAT
ne capture PAS entièrement. Mais attention :
- Seulement 156/920 tools (17%) sont connectés à des caps → Jaccard = 0 pour 83% des outils
- 100% des paires avec Jaccard > 0 ont aussi cosine > 0.5 → SHGAT est déjà bon là où Jaccard aide
- 282 traces / 158 transitions = échantillon bigram petit

**Verdict** : les structural features apportent un complément marginal à SHGAT.
Un A/B `NO_STRUCTURAL=true` tranchera si ce complément vaut le bruit qu'il ajoute.